File: android/app/src/main/kotlin/com/fpt/yuyama/MsiScannerChannel.kt
package com.fpt.yuyama

import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class MsiScannerChannel private constructor(
    flutterEngine: FlutterEngine
) : MethodChannel.MethodCallHandler {

    companion object {
        private const val CHANNEL = "com.fpt.yuyama/msi_scanner"

        fun registerWith(flutterEngine: FlutterEngine): MsiScannerChannel {
            return MsiScannerChannel(flutterEngine)
        }
    }

    private val channel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL)
    private val backgroundExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "decodeMsiYuv" -> decodeYuv(call, result)
            "decodeMsiBitmap" -> decodeBitmap(call, result)
            else -> result.notImplemented()
        }
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        backgroundExecutor.shutdown()
    }

    private fun decodeYuv(call: MethodCall, result: MethodChannel.Result) {
        val bytes = call.argument<ByteArray>("imageBytes")
        val width = call.argument<Int>("imageWidth") ?: 0
        val height = call.argument<Int>("imageHeight") ?: 0
        val rowStride = call.argument<Int>("rowStride") ?: width
        val checksumScheme = checksumSchemeFrom(call.argument<String>("checksumScheme"))

        if (bytes == null || width <= 0 || height <= 0 || rowStride <= 0) {
            result.error(
                "INVALID_ARGUMENT",
                "YUV luminance data, width, height, and rowStride are required",
                null
            )
            return
        }

        executeDecode(result, source = "yuv") {
            MsiNativeDecoder.decodeYuvLuminance(
                yArray = bytes,
                width = width,
                height = height,
                rowStride = rowStride,
                scheme = checksumScheme
            )
        }
    }

    private fun decodeBitmap(call: MethodCall, result: MethodChannel.Result) {
        val bytes = call.argument<ByteArray>("imageBytes")
        val checksumScheme = checksumSchemeFrom(call.argument<String>("checksumScheme"))

        if (bytes == null || bytes.isEmpty()) {
            result.error("INVALID_ARGUMENT", "Encoded image bytes are required", null)
            return
        }

        executeDecode(result, source = "bitmap") {
            val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            if (bitmap == null) null else MsiNativeDecoder.decodeBitmap(bitmap, checksumScheme)
        }
    }

    private fun executeDecode(
        result: MethodChannel.Result,
        source: String,
        decode: () -> String?
    ) {
        backgroundExecutor.execute {
            val startedAt = System.nanoTime()
            try {
                val decodedText = decode()
                val durationMs = ((System.nanoTime() - startedAt) / 1_000_000L).toInt()
                mainHandler.post {
                    result.success(
                        mapOf(
                            "text" to decodedText,
                            "durationMs" to durationMs,
                            "source" to source
                        )
                    )
                }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error(
                        "MSI_DECODE_FAILED",
                        e.message ?: "MSI decode failed",
                        null
                    )
                }
            }
        }
    }

    private fun checksumSchemeFrom(value: String?): MsiNativeDecoder.ChecksumScheme {
        return try {
            MsiNativeDecoder.ChecksumScheme.valueOf(value ?: "AUTO")
        } catch (ignored: IllegalArgumentException) {
            MsiNativeDecoder.ChecksumScheme.AUTO
        }
    }
}

File: ios/Runner/MsiScannerChannel.swift
import Flutter
import Foundation

final class MsiScannerChannel {
  private static let channelName = "com.fpt.yuyama/msi_scanner"

  private let channel: FlutterMethodChannel
  private let decodeQueue = DispatchQueue(label: "com.fpt.yuyama.msi_scanner.decode")

  init(binaryMessenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(
      name: Self.channelName,
      binaryMessenger: binaryMessenger
    )
    channel.setMethodCallHandler(handle)
  }

  func dispose() {
    channel.setMethodCallHandler(nil)
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "decodeMsiYuv":
      decodeYuv(call: call, result: result)
    case "decodeMsiBitmap":
      decodeBitmap(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func decodeYuv(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
      let bytes = arguments["imageBytes"] as? FlutterStandardTypedData,
      let width = arguments["imageWidth"] as? Int,
      let height = arguments["imageHeight"] as? Int
    else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENT",
          message: "YUV luminance data, width, and height are required",
          details: nil
        )
      )
      return
    }

    let rowStride = arguments["rowStride"] as? Int ?? width
    let checksumScheme = checksumScheme(from: arguments["checksumScheme"] as? String)

    guard width > 0, height > 0, rowStride > 0 else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENT",
          message: "YUV width, height, and rowStride must be greater than zero",
          details: nil
        )
      )
      return
    }

    executeDecode(result: result, source: "yuv") {
      MsiNativeDecoder.decodeYuvLuminance(
        bytes,
        width: width,
        height: height,
        rowStride: rowStride,
        scheme: checksumScheme
      )
    }
  }

  private func decodeBitmap(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
      let bytes = arguments["imageBytes"] as? FlutterStandardTypedData,
      !bytes.data.isEmpty
    else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENT",
          message: "Encoded image bytes are required",
          details: nil
        )
      )
      return
    }

    let checksumScheme = checksumScheme(from: arguments["checksumScheme"] as? String)

    executeDecode(result: result, source: "bitmap") {
      MsiNativeDecoder.decodeImageData(bytes, scheme: checksumScheme)
    }
  }

  private func executeDecode(
    result: @escaping FlutterResult,
    source: String,
    decode: @escaping () -> String?
  ) {
    decodeQueue.async {
      let startedAt = DispatchTime.now().uptimeNanoseconds
      let decodedText = decode()
      let durationMs = Int((DispatchTime.now().uptimeNanoseconds - startedAt) / 1_000_000)

      DispatchQueue.main.async {
        result([
          "text": decodedText as Any,
          "durationMs": durationMs,
          "source": source,
        ])
      }
    }
  }

  private func checksumScheme(from value: String?) -> MsiChecksumScheme {
    guard let value = value else { return .auto }
    return MsiChecksumScheme(rawValue: value) ?? .auto
  }
}
